codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttMessageType.java

```diff
@@ -46,12 +63,10 @@ public int value() {
     }
 
     public static MqttMessageType valueOf(int type) {
```

Comment:
What about simple switch statement? In that case we don't need to create and hold static variable.

Reply:
I see some reasons, the switch statement is:

1. tied to the encoding of MqttMessageType
2. is likely using many branches that the JVM will try to order accordly to some policies (for no reasons, given that this is for pure encoding reasons)
3. has a bigger byte code size, reducing the chances to be inlined or that its caller will be inlined when compiled into an bigger method

In general I see that the lookup tables perform a bit better for cases like these, but I haven't produced a reliable bench yet for this one to be 100% sure

Reply:
I see from http://hg.openjdk.java.net/jdk8/jdk8/hotspot/rev/34bd5e86aadb that C2 should generate a jump table in ASM here due to the number of choices, saving the binary jump table approach used otherwise: I admit I am curious to check how it perform va this PR...

Reply:
@franz1981 so you want to pull this in or do some more experiments first ? I am cool with both btw

Reply:
My test shows that switch is faster rather than static field array field access. However, your comment regarding byte code may be valid. So you need to check that with full decoder flow to prove that.

Reply:
Let me try writing a small bench here hud: I see other points on MQTT where enums valueOf lookup could be simplified like this and it's used on the hot paths.

Reply:
@doom369

Re 

> My test shows that switch is faster rather than static field array access

That surprise me: the pattern of message types you've used to perform the lookup is long enough to trick the branch predictor?
I see that can make a lot of difference while dealing with real use case vs synthetic workload

> However, your comment regarding byte code size may be valid

Yep, I have been recently bitten by inlining budget/threshold exhaustion with Netty (see https://github.com/netty/netty/pull/10368)

Reply:
@franz1981 well, I didn't dig into root cause. Here is my benchmark:

```
@BenchmarkMode(Mode.AverageTime)
@Fork(1)
@State(Scope.Thread)
@Warmup(iterations = 5, time = 1)
@OutputTimeUnit(TimeUnit.NANOSECONDS)
@Measurement(iterations = 10, time = 1)
public class ArrayVsSwitch {

    @Param({"1", "14"})
    int type;

    @Benchmark
    public MqttMessageType getViaArray() {
        return MqttMessageType.valueOf(type);
    }

    @Benchmark
    public MqttMessageType getViaSwitch() {
        switch (type) {
            case 1 :
                return MqttMessageType.CONNECT;
            case 2 :
                return MqttMessageType.CONNACK;
            case 3 :
                return MqttMessageType.PUBLISH;
            case 4 :
                return MqttMessageType.PUBACK;
            case 5 :
                return MqttMessageType.PUBREC;
            case 6 :
                return MqttMessageType.PUBREL;
            case 7 :
                return MqttMessageType.PUBCOMP;
            case 8 :
                return MqttMessageType.SUBSCRIBE;
            case 9 :
                return MqttMessageType.SUBACK;
            case 10 :
                return MqttMessageType.UNSUBSCRIBE;
            case 11 :
                return MqttMessageType.UNSUBACK;
            case 12 :
                return MqttMessageType.PINGREQ;
            case 13 :
                return MqttMessageType.PINGRESP;
            case 14 :
                return MqttMessageType.DISCONNECT;
            default:
                throw new IllegalArgumentException("unknown message type: " + type);
        }
    }

    public enum MqttMessageType {
        CONNECT(1),
        CONNACK(2),
        PUBLISH(3),
        PUBACK(4),
        PUBREC(5),
        PUBREL(6),
        PUBCOMP(7),
        SUBSCRIBE(8),
        SUBACK(9),
        UNSUBSCRIBE(10),
        UNSUBACK(11),
        PINGREQ(12),
        PINGRESP(13),
        DISCONNECT(14);

        private static final MqttMessageType[] VALUES;

        static {
            // this prevent values to be assigned with the wrong order
            // and ensure valueOf to work fine
            MqttMessageType[] values = values();
            VALUES = new MqttMessageType[values.length + 1];
            for (MqttMessageType mqttMessageType : values) {
                int value = mqttMessageType.value;
                if (VALUES[value] != null) {
                    throw new AssertionError("value already in use: " + value);
                }
                VALUES[value] = mqttMessageType;
            }
        }

        private final int value;

        MqttMessageType(int value) {
            this.value = value;
        }

        public int value() {
            return value;
        }

        public static MqttMessageType valueOf(int type) {
            if (type <= 0 || type >= VALUES.length) {
                throw new IllegalArgumentException("unknown message type: " + type);
            }
            return VALUES[type];
        }
    }
}
```

Result:

```
Benchmark                   (type)  Mode  Cnt  Score   Error  Units
ArrayVsSwitch.getViaArray        1  avgt   10  3.464 ± 0.012  ns/op
ArrayVsSwitch.getViaArray       14  avgt   10  3.466 ± 0.012  ns/op
ArrayVsSwitch.getViaSwitch       1  avgt   10  3.169 ± 0.022  ns/op
ArrayVsSwitch.getViaSwitch      14  avgt   10  3.174 ± 0.044  ns/op
```

P. S. getViaSwitch is not wrapped into a method, but `valueOf` inlined as well. So it doesn't matter for that test.